pkg/git/git.go

```diff
@@ -0,0 +1,73 @@
+// Copyright The Shipwright Contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package git
+
+import (
+	"context"
+	"fmt"
+	"github.com/go-git/go-git/v5/config"
+	"github.com/go-git/go-git/v5/plumbing/transport"
+	"github.com/go-git/go-git/v5/storage/memory"
+	"regexp"
+
+	gogitv5 "github.com/go-git/go-git/v5"
+)
+
+const (
+	defaultRemote = "origin"
+	privateRepo   = "private"
+	publicRepo    = "public"
+	invalidRepo   = "invalid"
+)
+
+var (
+	// httpURLRegexFormat and sshURLRegexFormat match an invalid url
+	httpURLRegexFormat = regexp.MustCompile(`^((http:\/\/)|(https:\/\/))`)
+	sshURLRegexFormat  = regexp.MustCompile(`(ssh:\/\/[\w\d\.]+|.+@?.+\..+:)(:[\d]+){0,1}\/*(.*)`)
+	// httpsPublicURLRegexFormat and sshPublicURLRegexFormat match pubic git repository url
+	httpsPublicURLRegexFormat = regexp.MustCompile(`(https:\/\/)?(github\.)+com(\/)`)
+	sshPublicURLRegexFormat   = regexp.MustCompile(`(git+@github.com:)(:[\d]+){0,1}/*(.*)`)
+)
+
+// ValidateGitURLExists validate if a source URL exists or not
+func ValidateGitURLExists(ctx context.Context, urlPath string) error {
+	repoKind := IdentifyGitURLType(urlPath)
+	switch repoKind {
+	case publicRepo:
+		repo := gogitv5.NewRemote(memory.NewStorage(), &config.RemoteConfig{
+			Name: defaultRemote,
+			URLs: []string{urlPath},
+		})
+		// Note: When the urlPath is an valid public path, however, this path doesn't exist, func will return `authentication required`, this maybe mislead
+		// So convert this error message to `repository not found`
+		_, err := repo.List(&gogitv5.ListOptions{})
+		if err != nil && err == transport.ErrAuthenticationRequired {
+			return fmt.Errorf("repository not found")
+		} else if err != nil {
+			return err
+		}
+		return nil
+	case invalidRepo:
+		return fmt.Errorf("invalid source url")
+	case privateRepo:
+		// Skip verify private repo
+		// TODO: check private repo in future
+		return nil
+	default:
+		return fmt.Errorf("unknown repository kind %s", urlPath)
+	}
+}
+
+// IdentifyGitURLType validates the given URL format is valid and public or not
+func IdentifyGitURLType(url string) string {
+	if sshPublicURLRegexFormat.MatchString(url) || httpsPublicURLRegexFormat.MatchString(url) {
```

Comment:
This could probably be nicer if you parse the url string into a *url.URL, and look at URL elements.

Reply:
Or probably better than this...

```go
import "github.com/go-git/go-git/v5/plumbing/transport"

endpoint, err :=  transport.NewEndpoint("git@github.com:shipwright-io/build.git")
```

Which gets you one of these https://pkg.go.dev/github.com/go-git/go-git/v5@v5.2.0/plumbing/transport#Endpoint

Reply:
If I understand clearly. Although I check the `Protocol ` of this endpoint, it can't tell me this repo exists or not. For example, check `https://github.ibm.com/shipwright-io/build` will return `https` protocol, but actually, this repo doesn't exist.

Reply:
@xiujuan95 this is replacing just the type detection logic. In case of http or https, you then run the logic in `ValidateGitURLExists` that you did so far in case you detected `public`.

Reply:
@SaschaSchwarze0 Ok, but in case of http or https, it doesn't mean this is a public repo, correct? However, the target of `IdentifyGitURLType ` function is to verify URL is a private or public URL.

Reply:
Let's go one step back @xiujuan95. What do we want to reach ? My understanding: we want to determine whether we will be able to clone the URL as part of a `git clone` once a `BuildRun` is created. This is unrelated to whether a repository is private or public. Git repository URLs can use HTTP(S) or SSH transport. As a first step, your code handles HTTP(S) transport because to test a SSH endpoint, it would need to check if a secret is defined on the source and extract and use the private key in there.

So, assuming we have a HTTP(S) URL, we will need to determine if we can reach. If we can reach it, we know that the `git clone` will also work. I am actually not sure if the opposite is also true, because if the repository is private, then we will fail here. But if there is a correct basic-auth secret associated in the Build, it will work.

Longer story, shorter conclusion:

1. Based on the URL, you cannot determine whether a repository is private or public.
2. For the full validation of the build's source, one will need to consider URL and associated secret.

Reply:
we dont want to validate repositories where an authentication is required for now,  the SSH is one of them.

Reply:
@bigkevmcd we adopted the `transport.NewEndpoint()` way, its better for the pr, thanks for the advise.